### PR TITLE
De-LLVM the unchecked shifts [MCP#693]

### PR DESCRIPTION
**Pull request by scottmcm**
This is just one part of the MCP (https://github.com/rust-lang/compiler-team/issues/693), but it's the one that IMHO removes the most noise from the standard library code.

Seems net simpler this way, since MIR already supported heterogeneous shifts anyway, and thus it's not more work for backends than before.

r? WaffleLapkin

### CODE REVIEW

File: library/core/src/intrinsics/mir.rs [hunk 1]
```diff
@@ -235,6 +235,7 @@
 //!  - Unary and binary operations use their normal Rust syntax - `a * b`, `!c`, etc.
 //!  - The binary operation `Offset` can be created via [`Offset`].
 //!  - Checked binary operations are represented by wrapping the associated binop in [`Checked`].
+//!  - Unchecked binary operations are represented by wrapping the associated binop in [`Unchecked`].
```
Comment (RalfJung) [inline library/core/src/intrinsics/mir.rs:238]: Can you add a test for this?

**Comment by RalfJung** [comment]
> As far as I can tell, CTFE, Miri, and Codegen all supported mixed unchecked shifts already without me doing anything. So this just gives a way to write tests exercising those.

This comment confusingly disappeared, but for CTFE/Miri it is true. The MIR binops support different types for LHS and RHS (and these intrinsics lower to MIR binops).

**Comment by WaffleLapkin** [comment]
> but it's the one that IMHO removes the most noise from the standard library code.

I was actually stuck on figuring the shifts out. Since you can have integer shifts with any integer rhs, backends will need to cast, no matter what, so I couldn't figure out how to make everything better for everyone.

I have all the other changes ready (I think), so I'll open a PR once we merge this (I'll try to review soon).

**Review by WaffleLapkin — CHANGES_REQUESTED** [review]
I have some small nits and am somewhat confused why we want to make intrinsics generic wrt lhs. But LGTM on the general approach of moving things from std to the backend.

This does not particularly help us at my dayjob (we would like `bx.shl` to take rhs which is `u32`, i.e. we would like to have a different `build_shift_expr_rhs`), but I _do_ like this change anyway. 😅

File: library/core/src/intrinsics.rs [hunk 2]
```diff
@@ -2224,18 +2224,20 @@ extern "rust-intrinsic" {
     /// Safe wrappers for this intrinsic are available on the integer
     /// primitives via the `checked_shl` method. For example,
     /// [`u32::checked_shl`]
+    #[cfg(not(bootstrap))]
     #[rustc_const_stable(feature = "const_int_unchecked", since = "1.40.0")]
     #[rustc_nounwind]
-    pub fn unchecked_shl<T: Copy>(x: T, y: T) -> T;
+    pub fn unchecked_shl<T: Copy, U: Copy>(x: T, y: U) -> T;
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply library/core/src/intrinsics.rs:2230]: A bit weird (given that we don't publicly expose this), but given this does not make backend more complex, ig this is fine.

**Review by WaffleLapkin — APPROVED** [review]

**Pull request merged**